src/common/com/intellij/plugins/haxe/util/HaxeResolveUtil.java

```diff
@@ -725,6 +710,36 @@ private static HaxeClassResolveResult getHaxeClassResolveResultInternal(@Nullabl
     return getHaxeClassResolveResult(initExpression, specialization);
   }
 
+  private static HaxeClassResolveResult searchForIterableTypeRecursively(HaxeClassResolveResult resolveResult) {
+    final HaxeClass resolveResultHaxeClass = resolveResult.getHaxeClass();
+    final HaxeGenericResolver resolver = resolveResult.getGenericResolver();
+    final HaxeGenericSpecialization resultSpecialization = resolveResult.getSpecialization();
+
+    // try next
+    HaxeClassResolveResult result = getResolveMethodReturnType(resolver, resolveResultHaxeClass, "next", resultSpecialization);
+    if (result.getHaxeClass() != null) {
+      return result;
+    }
+    // try iterator
+    HaxeClassResolveResult iteratorResult =
+      getResolveMethodReturnType(resolver, resolveResultHaxeClass, "iterator",
+                                 resultSpecialization.getInnerSpecialization(resolveResultHaxeClass));
+
+    HaxeClass iteratorResultHaxeClass = iteratorResult.getHaxeClass();
+    // Now, look for iterator's next
+    result = getResolveMethodReturnType(resolver, iteratorResultHaxeClass, "next", iteratorResult.getSpecialization());
+
+
+    if (result.getHaxeClass() == null) {
+      // check underlying types
+      SpecificHaxeClassReference underlyingClassReference = resolveResultHaxeClass.getModel().getUnderlyingClassReference(resolver);
+      if(underlyingClassReference != null) {
+        result = searchForIterableTypeRecursively(underlyingClassReference.asResolveResult());
```

Comment:
I think we need circular reference protection on this one, because the user can make two classes inherit from each other in code even though it won't compile.